### ISSUE TITLE
PhysicalBones broke themselves

### ISSUE THREAD

**Issue by Favkis** [godotengine/godot#40549]
**Godot version:**
steam stable 3.2

**OS/device including version:**
windows 10

**Issue description:**
Physboxes broke themselves

**Steps to reproduce:**
no idea, I've been doing my game and now all physboxes are broken 
![image](https://user-images.githubusercontent.com/46406204/87974818-4f907980-cad3-11ea-8fff-708f7a1fc7c1.png)

In editor they are where they should be
![image](https://user-images.githubusercontent.com/46406204/87974861-620ab300-cad3-11ea-8e7e-e46cfca79ef8.png)
![image](https://user-images.githubusercontent.com/46406204/87974897-73ec5600-cad3-11ea-9c33-5b8d1f44a698.png)

Hitboxes on the other hand (Boneattachments with staticbody inside and collisionshape) are still where they should be, only all PhysicalBones broke themselves.

**Comment by Calinou** [comment]
@Favkis Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

**Comment by Favkis** [comment]
> @Favkis Please upload a [minimal reproduction project](https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#provide-a-simple-example-project) to make this easier to troubleshoot.

I don't know how. I just reported it as a bug, I cannot reproduce it. I just had bot.tscn that contains script on them and human.tscn inside. And the one in scene had his PhysicalBones broken after some time, while BoneAttachments were fine.
Sorry I cannot give more info or reproduce.

**Comment by Calinou** [comment]
@Favkis Bug reports that are difficult to reproduce are unfortunately not very useful for us. I'll have to close this one in 7 days if nobody else manages to reproduce it.

**Comment by Favkis** [comment]
I have this bug again. This time whole class broke their physboxes. I've changed root type from KinematicBody to Human (my class) and my physboxes are permanently stuck in animation frame 
![image](https://user-images.githubusercontent.com/46406204/88081355-eec37880-cb88-11ea-992b-4976c83c5de7.png)

**Comment by Favkis** [comment]
> @Favkis Bug reports that are difficult to reproduce are unfortunately not very useful for us. I'll have to close this one in 7 days if nobody else manages to reproduce it.

I have this bug again. This time whole class broke their physboxes. I've changed root type from KinematicBody to Human (my class) and my physboxes are permanently stuck in animation frame
![image](https://user-images.githubusercontent.com/46406204/88081401-000c8500-cb89-11ea-9119-b4c06a6ff354.png)

**Comment by Favkis** [comment]
is there any way to bake physbones position to rest animation again? It happens almsot everyday, almost everyday physical bones are baking into some animation frame form, so maybe I am pressing something? I hope so much, I hate redoing them every day

**Comment by Favkis** [comment]
Ok I've made workaround:
the problem is that PhysicalBone for some reason bakes body_offset from random animation frame:
![image](https://user-images.githubusercontent.com/46406204/88084544-419f2f00-cb8d-11ea-90e3-18d68b481fc9.png)
The workaround is to do this:
`func fix_physical_bones():
	var translate=Transform(\
		Vector3(1,0,0),\
		Vector3(0,1,0),\
		Vector3(0,0,1),\
		Vector3(0,0,0))
	for bone in ($Armature/Skeleton.get_children()):
		if bone is PhysicalBone:
			bone.body_offset=translate`

**Comment by Favkis** [comment]
it's getting rediculous, workaround doesn't work ANYMORE. Literally it was working when I wrote previous test, not it DOESN'T. What the fuck is going on with these PhysicalBones???

**Comment by Favkis** [comment]
Workaround #2:
Set up Body Offset by your hand to default matrix values
1 0 0
0 1 0
0 0 1
0 0 0
DO NOT TURN ON AnimationTree in your character scene, only do so in game with script or you risk breaking Body Offset and will have to fix it by hand again because in game it doesn't work.
I would also suggest to not play animations in Animation Player, or at least never leave in non idle (default imported) animation your character tab.

**Comment by marstaik** [comment]
I think I know whats going on, its happening to me as well. I am investigating.

**Comment by marstaik** [comment]
There is some editor only code that updates the body offsets when the transform changes.

When an animation that is played back in the editor with animate physical bones enabled on the skeleton, it causes the body offsets to be recomputed.

I need to figure out how to differentiate between a transform changed by the editor versus an actual transform change...

**Comment by nicohvi** [comment]
I'd just like to add that this happened to me as well (10 months later), and it wasn't very straightforward to find out what the problem was. Perhaps add something about this shortcoming in the documentation?

**Comment by Calinou** [comment]
> I'd just like to add that this happened to me as well (10 months later), and it wasn't very straightforward to find out what the problem was. Perhaps add something about this shortcoming in the documentation?

I don't know what exactly should be pinpointed here, so feel free to [open a pull request on the class reference XML](https://docs.godotengine.org/en/latest/community/contributing/updating_the_class_reference.html) to document this :slightly_smiling_face:

**Comment by nicohvi** [comment]
Okay @Calinou, I'll do that! 👍 

I'd like to add that I'm really enjoying working with the engine, thanks for all the great work you guys are doing 😃

**Comment by Calinou** [comment]
@nicohvi Bump – if you have time, remember to update the class reference as I mentioned above.

**Comment by Zireael07** [comment]
Looks like this is missing the documentation label/tag then?

**Comment by Favkis** [comment]
oh my fucking god please for the love of god somebody fix it plesae i beg you i have to manually reset all physboxes again oh my god
![image](https://user-images.githubusercontent.com/46406204/210084982-a4afc1c1-15d9-4e99-8ac6-1769ef6b1146.png)

**Comment by Favkis** [comment]
im sorry it just takes ages to undo and this bug keeps hunting me everywhere my god

**Comment by Calinou** [comment]
We still need a minimal reproduction project, as recreating a PhysicalBone setup from scratch takes a lot of time.

**Comment by Qoltaic** [comment]
@Calinou I have came across what I think is the exact issue, so I will put it here along with a reproduction project.

**Description:**
The issue occurs when the scene is saved after playing an animation and stopping in a pose that transforms the bone the physicalbone3d is attached to. Once the scene is closed and reopened, the physicalbone3d will be at the paused location it was saved at relative to the rest pose. The behavior can be repeated many times and will result in an additive behavior. I have included a video as well.

**Reproduction Steps:**
1_ In the reproduction file, open the "Character" scene and select the animation player and press the play button and pause the animation in the middle.
2_ Save the scene and close it.
3_ Open the character scene again. 

https://github.com/godotengine/godot/assets/146401461/cd620027-80e0-45d2-8d69-64cce05b170d


[Reproduction(PhysicalBone3d).zip](https://github.com/godotengine/godot/files/12755678/Reproduction.PhysicalBone3d.zip)

**Comment by levidavidmurray** [comment]
![image](https://github.com/godotengine/godot/assets/46084870/651c51d1-4b4c-4ce0-b8b1-19578333c417)

The first time I ran into this was after setting up my physical bones for the first time—before I had committed... Since then I've made a point to always set the skeleton to `show_rest_only` before saving. Sometimes I slip up and forget, but I'm usually able to revert the diffs easy enough. Still very frustrating.

What's the future of all this skeleton, physical bone, IK stuff? I thought I remember reading somewhere that there was a rework either in progress, or on the roadmap, but that may have been specifically the whole IK system.

So, is this specific issue worth looking into, or will everything around it change in coming versions of Godot?

**Comment by ThomOr** [comment]
Ran into this issue recently. It's kind of a showstopper for me, as there's no easy way to reset the bone positions that I can see.

**Comment by bergvall95** [comment]
Running into this as well: 
![image](https://github.com/godotengine/godot/assets/36799924/b6801d2e-e24c-4a5b-8d6f-63aede94bc7f)


https://github.com/godotengine/godot/assets/36799924/d424fb6d-d8ce-4ddd-bb5c-1e88b1889957

I don't exactly have repro steps, but the project is public and on GH:  https://github.com/bergvall95/godot-game

**Comment by bergvall95** [comment]
Okay after messing with it for a while I do have repro steps:  
(I'm a bit of a noob in godot so bare with me) 
1. Have skeleton with Bones 
2.  AnimationPlayer set Active => off
3.  Animation tree set to inactive
4. select animation tab
5. switch between different animations


There seem to be several routes to cause this bug to happen, likely that it is multiple bugs I would imagine. 
Another case is the same as above, but instead of just switching an animation, pressing save during an animation can mess them up. 

https://github.com/godotengine/godot/assets/36799924/c08123ae-6ad1-4816-a22c-bc330b6a0b6a

**Comment by quadrillegams** [comment]
Still broken in 4.4 beta 2. I'm curious as to why this issue is tracked under GodotPhysics3D issues, as the underlying problems have nothing to do with GodotPhysics itself, and everything to do with the interaction between Skeleton3D, active AnimationTrees, and active AnimationPlayers.

**Comment by HungryHedgehog** [comment]
Yeah still broken have to fix it for the 10th time now, glad to at least know why it happens. I need to work on the animations so I can't just turn off the animations in the editor. At first I thought it was the blender import but it breaks even if I don't import anything. It doesnt just mess with the offset but also with the transforms of the colliders in the bones for some reason.
My current work around is to copy over the skeleton form another scene and then change the few parts that need to be tweaked, but it's really not ideal having to constantly fix all the hitboxes of my character because I touched the animation nodes.
